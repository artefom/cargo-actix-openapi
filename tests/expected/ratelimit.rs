//! API auto-generated by apigen

use std::fmt::Display;

use std::{collections::HashMap, fmt::Debug};

use serde::{Deserialize, Serialize};

use actix_web::{
    get, http::StatusCode, middleware::NormalizePath, web, App, HttpRequest, HttpResponse,
    HttpServer, ResponseError,
};

use async_trait::async_trait;

// Defaults
// -------------------------------


// Enums
// -------------------------------


// Struct
// -------------------------------

#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct QuotaDetailsPath {
    pub quota: String,
}

/// Quota specification
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct Quota {
    pub replanish_interval: f64,
    pub burst_capacity: i64,
}

#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct MatchRule {
}

/// State information of the quota
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct QuotaState {
    pub earliest_next_available: f64,
    pub remaining_burst_capacity: i64,
}

/// Quota statistics, purely descriptive. Not used in Rate limiting decisions.
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct QuotaStats {
    pub rpm: i64,
}

/// Full information about quota
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct QuotaDetails {
    pub quota: Quota,
    #[serde(rename="match")]
    pub match_: Vec<MatchRule>,
    pub state: QuotaState,
    pub stats: QuotaStats,
}

#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct MatchRule {
}

#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct CellTestQuery {
    pub query: MatchRule,
}

/// Quota specification
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct Quota {
    pub replanish_interval: f64,
    pub burst_capacity: i64,
}

/// State information of the quota
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct QuotaState {
    pub earliest_next_available: f64,
    pub remaining_burst_capacity: i64,
}

/// Information about current cell state and matched quotas.
/// Matched quotas are computed based on query.
/// Info and state are computed dynamically based on matched quotas.
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct CellDetails {
    pub quotas: Vec<String>,
    pub info: Quota,
    pub state: QuotaState,
}

#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct MatchRule {
}

#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct CellUpdateQuery {
    pub query: MatchRule,
}

/// Quota specification
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct Quota {
    pub replanish_interval: f64,
    pub burst_capacity: i64,
}

/// State information of the quota
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct QuotaState {
    pub earliest_next_available: f64,
    pub remaining_burst_capacity: i64,
}

/// Information about current cell state.
/// Info and state are computed dynamically based on matched quotas.
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct CellInfo {
    pub info: Quota,
    pub state: QuotaState,
}

/// Result of the cell update. Allowed/Denied flag + cell info
#[derive(Serialize, Deserialize, Clone, PartialEq)]
pub struct UpdateResult {
    pub allowed: bool,
    pub details: CellInfo,
}


// Error with details
// -------------------------------

/// Bails with detailed api error
#[macro_export]
macro_rules! apibail {
    ($err:expr,$msg:expr) => {
        return Err($crate::server::api::Detailed {
            error: $err,
            details: $msg.to_string(),
        })
    };
}

pub trait StatusCoded {
    fn status_code(&self) -> StatusCode;
}

#[derive(Debug)]
pub struct Detailed<E> {
    pub error: E,
    pub details: String,
}

impl<E: Display> Display for Detailed<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}. Reason: {}", self.error, self.details)
    }
}

impl<E: Display + Debug> std::error::Error for Detailed<E> {}

impl<E: Display + Debug> ResponseError for Detailed<E>
where
    E: StatusCoded,
{
    fn status_code(&self) -> StatusCode {
        self.error.status_code()
    }
}

/// Converts some result to Result with detailed api error
pub trait ApiErr<T, E> {
    /// Wrap the error value with additional context.
    fn apierr<C>(self, err: C) -> Result<T, Detailed<C>>
    where
        C: Display + Send + Sync + 'static;
}

impl<T, E> ApiErr<T, E> for Result<T, E>
where
    E: Debug + Send + Sync + 'static,
{
    fn apierr<C>(self, err: C) -> Result<T, Detailed<C>>
    where
        C: Display + Send + Sync + 'static,
    {
        // Not using map_err to save 2 useless frames off the captured backtrace
        // in ext_context.
        match self {
            Ok(ok) => Ok(ok),
            Err(original_error) => Err(Detailed {
                error: err,
                details: format!("{:?}", original_error),
            }),
        }
    }
}


// Error
// -------------------------------

/// Status NOT_FOUND:
/// Quota not found
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub enum QuotaDetailsResponseError {
    QuotaNotFound,
}

impl Display for QuotaDetailsResponseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::QuotaNotFound => {
                write!(f, "Quota not found")
            },
        }
    }
}

impl StatusCoded for QuotaDetailsResponseError {
    fn status_code(&self) -> StatusCode {
        match self {
            Self::QuotaNotFound => StatusCode::NOT_FOUND,
        }
    }
}

/// Status BAD_REQUEST:
/// Duplicate key in query
/// 
/// Status NOT_FOUND:
/// No quotas matching given query found
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub enum CellTestResponseError {
    DuplicateQueryKey,
    NoQuotasMatchingQueryFound,
}

impl Display for CellTestResponseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::DuplicateQueryKey => {
                write!(f, "Duplicate query key")
            },
            Self::NoQuotasMatchingQueryFound => {
                write!(f, "No quotas matching query found")
            },
        }
    }
}

impl StatusCoded for CellTestResponseError {
    fn status_code(&self) -> StatusCode {
        match self {
            Self::DuplicateQueryKey => StatusCode::BAD_REQUEST,
            Self::NoQuotasMatchingQueryFound => StatusCode::NOT_FOUND,
        }
    }
}

/// Status BAD_REQUEST:
/// Duplicate key in query
/// 
/// Status NOT_FOUND:
/// No quotas matching given query found
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub enum CellUpdateResponseError {
    DuplicateQueryKey,
    NoQuotasMatchingQueryFound,
}

impl Display for CellUpdateResponseError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::DuplicateQueryKey => {
                write!(f, "Duplicate query key")
            },
            Self::NoQuotasMatchingQueryFound => {
                write!(f, "No quotas matching query found")
            },
        }
    }
}

impl StatusCoded for CellUpdateResponseError {
    fn status_code(&self) -> StatusCode {
        match self {
            Self::DuplicateQueryKey => StatusCode::BAD_REQUEST,
            Self::NoQuotasMatchingQueryFound => StatusCode::NOT_FOUND,
        }
    }
}

// Api service
// -------------------------------


#[async_trait(?Send)]
pub trait ApiService<S>
where
    S: Send + Sync + 'static,
{
    async fn health(
        data: web::Data<S>,
    ) -> web::Json<String>;
    async fn quota_list(
        data: web::Data<S>,
    ) -> web::Json<Vec<String>>;
    async fn quota_details(
        data: web::Data<S>,
    ) -> Result<web::Json<QuotaDetails>,QuotaDetailsResponseError>;
    async fn cell_test(
        data: web::Data<S>,
    ) -> Result<web::Json<CellDetails>,CellTestResponseError>;
    async fn cell_update(
        data: web::Data<S>,
    ) -> Result<web::Json<UpdateResult>,CellUpdateResponseError>;
}


// Run service function (+ helper functions)
// -----------------------------------------

static OPENAPI_FILE: &'static str = include_str!("static/openapi.yaml");
static DOCS_PAGE: &'static str = include_str!("static/docs.html");

/// Documentation
#[get("/openapi.yaml")]
async fn openapi() -> String {
    OPENAPI_FILE.to_string()
}

/// Documentation
#[get("/docs")]
async fn docs() -> HttpResponse {
    HttpResponse::build(StatusCode::OK)
        .content_type("text/html; charset=utf-8")
        .body(DOCS_PAGE)
}

#[get("/")]
async fn redirect_to_docs() -> HttpResponse {
    HttpResponse::build(StatusCode::PERMANENT_REDIRECT)
        .append_header(("Location", "docs"))
        .body("")
}

pub async fn run_service<T, S>(bind: &str, initial_state: S) -> Result<(), std::io::Error>
where
    T: ApiService<S> + 'static,
    S: Send + Sync + 'static,
{
    let app_data = web::Data::new(initial_state);

    HttpServer::new(move || {
        App::new()
            .app_data(app_data.clone())
            .wrap(NormalizePath::trim())
            .service(openapi)
            .service(redirect_to_docs)
            .service(docs)
            .route(
                "/health",
                web::get().to(T::health)
            )
            .route(
                "/quota",
                web::get().to(T::quota_list)
            )
            .route(
                "/quota/{quota}",
                web::get().to(T::quota_details)
            )
            .route(
                "/cell/test",
                web::get().to(T::cell_test)
            )
            .route(
                "/cell/update",
                web::post().to(T::cell_update)
            )
    })
    .bind(bind)?
    .run()
    .await?;

    Ok(())
}