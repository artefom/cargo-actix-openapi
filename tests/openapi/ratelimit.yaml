---
openapi: 3.0.3
info:
  title: ratelimit
  version: 1.0.0
  description: |
    Limits rates (with Rust)
    Implements Leaky bucket using Generic cell rate algorithm (GCRA).
    Allows for request bursts.
servers:
  - url: ./
paths:
  '/health':
    get:
      tags:
        - health
      summary: Check service health
      operationId: health
      responses:
        '200':
          description: 'Service is ready'
          content:
            application/json:
              schema:
                type: string
                default: 'Ok'
  '/quota':
    get:
      tags:
        - quota
      summary: List quotas
      operationId: quota_list
      responses:
        '200':
          description: List of all quotas
          content:
            application/json:
              schema:
                title: Response Quota List Quota Get
                type: array
                items:
                  type: string
                example:
                  - megb
                  - testquota
                  - anotherquota
  '/quota/{quota}':
    get:
      tags:
        - quota
      summary: Get quota details
      operationId: quota_details
      parameters:
        - $ref: '#/components/parameters/quota'
      responses:
        '200':
          description: Quota details, including current rpm and GCRA state
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/QuotaDetails'
        '404':
          description: Quota not found
          content:
            application/json:
              schema:
                title: Details
                type: string
                enum: ['Quota not found']
  '/cell/test':
    get:
      tags:
        - rate limiting
      summary: Get current rate limitation state for given query
      description: |
        Allows to probe for rate limits without updating quota states.
        Usefull for monitoring.
      operationId: cell_test
      parameters:
        - $ref: '#/components/parameters/query'
      responses:
        '200':
          description: Successful Response
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/CellDetails'
        '400':
          description: Duplicate key in query
          content:
            application/json:
              schema:
                title: Details
                type: string
                enum: ['Duplicate query key']
        '404':
          description: No quotas matching given query found
          content:
            application/json:
              schema:
                title: Details
                type: string
                enum: ['No quotas matching query found']
  '/cell/update':
    post:
      tags:
        - rate limiting
      summary: Try to accomodate for one request
      description: |
        Will update states of all quotas that match given query.
        Strictest rate limitation of all matched queries applies
      operationId: cell_update
      parameters:
        - $ref: '#/components/parameters/query'
      responses:
        '200':
          description: Successful Response
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/UpdateResult'
        '400':
          description: Duplicate key in query
          content:
            application/json:
              schema:
                title: Details
                type: string
                enum: ['Duplicate query key']
        '404':
          description: No quotas matching given query found
          content:
            application/json:
              schema:
                title: Details
                type: string
                enum: ['No quotas matching query found']
components:
  parameters:
    quota:
      name: quota
      required: true
      description: Quota label - Unique quota identifier
      in: path
      schema:
        title: Quota
        type: string
        example: testquota
    query:
      name: query
      description: |
        Query that will be matched against quotas

        # Matching rules
        ---------------

        Quota matches the query if at least one of it's predicates (match section) matches the query.
        Predicate matches query if all its key/values are present and match key/values of the request query.
        If query key is not present in the predicate, it is disregarded.

        ## Example:

        given predicate:


        `
        {
          'carrier': 'MEGB'
          'endpoint': 'location'
        }
        `

        match results on queries:


        `?carrier=MEGB&endpoint=locations` - OK


        `?carrier=MEGB` - No match


        `?carrier=MEGB&endpoint=locations&sender=retailer-api` - OK


        `?carrier=MEGB&sender=retailer-api` - No match


        `?sender=retailer-api` - No match

      required: true
      in: query
      schema:
        $ref: '#/components/schemas/MatchRule'
  schemas:
    MatchRule:
      type: object
      title: MatchRule
      additionalProperties:
        type: string
      example:
        agency: testagency
        carrier: TEST
        endpoint: someendpoint
    MatchRules:
      type: array
      title: MatchRules
      description: |
        Collection of predicates to test agains incomming queries
        If at least one predicate is matching the incomming query, the rate limit is applied to the request
        Multiple rate limits can be applied to incomming request at once
      items:
        $ref: '#/components/schemas/MatchRule'
    Quota:
      type: object
      title: Quota
      description: Quota specification
      required:
        - replanish_interval
        - burst_capacity
      properties:
        replanish_interval:
          title: Replanish Interval
          type: number
          description: |
            The 'weight' of a single cell in milliseconds or emission interval.
            Maximum allowed requests per minute can be calculated as: 60 * 1000 / replanish_interval
            Controls sustainable Cell Rate
          example: 1000.0
        burst_capacity:
          title: Burst Capacity
          type: integer
          description: |
            Number of sequential cells allowed in a single burst
            A burst or clump of packets can arrive at a higher rate than determined by the emission interval
            In case there is unused burst capacity, quota can also exceed RPM in certain time frames.
            Burst capacity of 0 ensure that RPM is never exceeded but introduces a lot of delay.
            Burst capacity does not affect Sustainable Cell Rate
          example: 5
    QuotaDetails:
      type: object
      title: QuotaDetails
      description: Full information about quota
      required:
        - quota
        - match
        - state
        - stats
      properties:
        quota:
          $ref: '#/components/schemas/Quota'
        match:
          $ref: '#/components/schemas/MatchRules'
        state:
          $ref: '#/components/schemas/QuotaState'
        stats:
          $ref: '#/components/schemas/QuotaStats'
    QuotaState:
      type: object
      title: QuotaState
      description: State information of the quota
      required:
        - earliest_next_available
        - remaining_burst_capacity
      properties:
        earliest_next_available:
          title: Earliest Next Available
          type: number
          description: Earliest delay in ms from now when next cell is available
          example: 566.66
        remaining_burst_capacity:
          title: remaining Burst Capacity
          type: integer
          description: Current remaining burst capacity
          example: 3
    QuotaStats:
      type: object
      title: QuotaStats
      description: Quota statistics, purely descriptive. Not used in Rate limiting decisions.
      required:
        - rpm
      properties:
        rpm:
          title: Rpm
          type: integer
          description: Number of requests in last 60 seconds
          example: 36
    QuotaUpdate:
      type: object
      title: QuotaUpdate
      description: Quota update schema
      required:
        - quota
        - match
      properties:
        quota:
          $ref: '#/components/schemas/Quota'
        match:
          $ref: '#/components/schemas/MatchRules'
    CellDetails:
      type: object
      title: CellDetails
      description: |
        Information about current cell state and matched quotas.
        Matched quotas are computed based on query.
        Info and state are computed dynamically based on matched quotas.
      required:
        - quotas
        - info
        - state
      properties:
        quotas:
          type: array
          title: Quotas
          description: Matched quotas
          items:
            type: string
            description: Quota label
        info:
          $ref: '#/components/schemas/Quota'
        state:
          $ref: '#/components/schemas/QuotaState'
    CellInfo:
      type: object
      title: CellInfo
      description: |
        Information about current cell state.
        Info and state are computed dynamically based on matched quotas.
      required:
        - info
        - state
      properties:
        info:
          $ref: '#/components/schemas/Quota'
        state:
          $ref: '#/components/schemas/QuotaState'
    UpdateResult:
      type: object
      title: UpdateResult
      description: |
        Result of the cell update. Allowed/Denied flag + cell info
      required:
        - allowed
        - details
      properties:
        allowed:
          type: boolean
          title: "Allowed"
          description: |
            Indicates if request was allowed
            If request was allowed, rate limit state was already updated to accomodate
            this request.
            If request was rejected, rate limit was not updated
        details:
          $ref: '#/components/schemas/CellInfo'
