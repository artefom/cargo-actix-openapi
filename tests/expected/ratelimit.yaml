api:
  definitions:
  - name: QuotaDetailsPath
    data: !Struct
      doc: null
      properties:
      - name: quota
        rename: quota
        default: null
        type_: String
  - name: Quota
    data: !Struct
      doc: Quota specification
      properties:
      - name: replanish_interval
        rename: replanish_interval
        default: null
        type_: f64
      - name: burst_capacity
        rename: burst_capacity
        default: null
        type_: i64
  - name: MatchRule
    data: !Struct
      doc: null
      properties: []
  - name: QuotaState
    data: !Struct
      doc: State information of the quota
      properties:
      - name: earliest_next_available
        rename: earliest_next_available
        default: null
        type_: f64
      - name: remaining_burst_capacity
        rename: remaining_burst_capacity
        default: null
        type_: i64
  - name: QuotaStats
    data: !Struct
      doc: Quota statistics, purely descriptive. Not used in Rate limiting decisions.
      properties:
      - name: rpm
        rename: rpm
        default: null
        type_: i64
  - name: QuotaDetails
    data: !Struct
      doc: Full information about quota
      properties:
      - name: quota
        rename: quota
        default: null
        type_: Quota
      - name: match_
        rename: match
        default: null
        type_: Vec<MatchRule>
      - name: state
        rename: state
        default: null
        type_: QuotaState
      - name: stats
        rename: stats
        default: null
        type_: QuotaStats
  - name: QuotaDetailsResponseError
    data: !ApiErr
      variants:
      - name: QuotaNotFound
        detail: Quota not found
        code: NOT_FOUND
  - name: MatchRule
    data: !Struct
      doc: null
      properties: []
  - name: CellTestQuery
    data: !Struct
      doc: null
      properties:
      - name: query
        rename: query
        default: null
        type_: MatchRule
  - name: Quota
    data: !Struct
      doc: Quota specification
      properties:
      - name: replanish_interval
        rename: replanish_interval
        default: null
        type_: f64
      - name: burst_capacity
        rename: burst_capacity
        default: null
        type_: i64
  - name: QuotaState
    data: !Struct
      doc: State information of the quota
      properties:
      - name: earliest_next_available
        rename: earliest_next_available
        default: null
        type_: f64
      - name: remaining_burst_capacity
        rename: remaining_burst_capacity
        default: null
        type_: i64
  - name: CellDetails
    data: !Struct
      doc: |
        Information about current cell state and matched quotas.
        Matched quotas are computed based on query.
        Info and state are computed dynamically based on matched quotas.
      properties:
      - name: quotas
        rename: quotas
        default: null
        type_: Vec<String>
      - name: info
        rename: info
        default: null
        type_: Quota
      - name: state
        rename: state
        default: null
        type_: QuotaState
  - name: CellTestResponseError
    data: !ApiErr
      variants:
      - name: DuplicateQueryKey
        detail: Duplicate query key
        code: BAD_REQUEST
      - name: NoQuotasMatchingQueryFound
        detail: No quotas matching query found
        code: NOT_FOUND
  - name: MatchRule
    data: !Struct
      doc: null
      properties: []
  - name: CellUpdateQuery
    data: !Struct
      doc: null
      properties:
      - name: query
        rename: query
        default: null
        type_: MatchRule
  - name: Quota
    data: !Struct
      doc: Quota specification
      properties:
      - name: replanish_interval
        rename: replanish_interval
        default: null
        type_: f64
      - name: burst_capacity
        rename: burst_capacity
        default: null
        type_: i64
  - name: QuotaState
    data: !Struct
      doc: State information of the quota
      properties:
      - name: earliest_next_available
        rename: earliest_next_available
        default: null
        type_: f64
      - name: remaining_burst_capacity
        rename: remaining_burst_capacity
        default: null
        type_: i64
  - name: CellInfo
    data: !Struct
      doc: |
        Information about current cell state.
        Info and state are computed dynamically based on matched quotas.
      properties:
      - name: info
        rename: info
        default: null
        type_: Quota
      - name: state
        rename: state
        default: null
        type_: QuotaState
  - name: UpdateResult
    data: !Struct
      doc: |
        Result of the cell update. Allowed/Denied flag + cell info
      properties:
      - name: allowed
        rename: allowed
        default: null
        type_: bool
      - name: details
        rename: details
        default: null
        type_: CellInfo
  - name: CellUpdateResponseError
    data: !ApiErr
      variants:
      - name: DuplicateQueryKey
        detail: Duplicate query key
        code: BAD_REQUEST
      - name: NoQuotasMatchingQueryFound
        detail: No quotas matching query found
        code: NOT_FOUND
  operations:
  - name: health
    path: /health
    method: Get
    doc: Check service health
    param_path: null
    param_query: null
    param_body: null
    response: web::Json<String>
  - name: quota_list
    path: /quota
    method: Get
    doc: List quotas
    param_path: null
    param_query: null
    param_body: null
    response: web::Json<Vec<String>>
  - name: quota_details
    path: /quota/{quota}
    method: Get
    doc: Get quota details
    param_path: web::Path<QuotaDetailsPath>
    param_query: null
    param_body: null
    response: Result<web::Json<QuotaDetails>,QuotaDetailsResponseError>
  - name: cell_test
    path: /cell/test
    method: Get
    doc: Get current rate limitation state for given query
    param_path: null
    param_query: web::Query<CellTestQuery>
    param_body: null
    response: Result<web::Json<CellDetails>,CellTestResponseError>
  - name: cell_update
    path: /cell/update
    method: Post
    doc: Try to accomodate for one request
    param_path: null
    param_query: web::Query<CellUpdateQuery>
    param_body: null
    response: Result<web::Json<UpdateResult>,CellUpdateResponseError>
